### PR TITLE
Update Jest and jsdom

### LINKED ISSUE #11036: Update Jest to v27 and jsdom to v18
I think these two updates are interdependent, so we should create a single commit that updates both. See https://github.com/mozilla/addons-frontend/pull/10957 and https://github.com/mozilla/addons-frontend/pull/10612

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #11036

### CODE REVIEW

**Comment by bobsilverberg** [comment]
@willdurand Combining these seemed to resolve all of the problems I was having with the `jsdom` update, although we're still manually importing `TextEncoder` and `TextDecoder` as I'm not sure what we can do about the whole `amo/window.js` thing.

Also, as discussed at https://github.com/mozilla/addons-frontend/pull/10612#issuecomment-974261600, I am left with one failing test that I am having trouble figuring out. The test is for [`DismissibleTextForm`](https://github.com/mozilla/addons-frontend/blob/master/tests/unit/amo/components/TestDismissibleTextForm.js#L121-L127), and it is trying to test that the `textarea` is focused upon mount. Unfortunately something broke with this update to Jest, so that whereas before `document.activeElement` was the `textarea`, now it seems to be the `body` element. So the issue doesn't seem to be with how to test whether an element is focused, but rather with the fact that the element we expect is no longer getting the focus during the test.

TBPH, if there isn't a simple solution to this, I don't think this test is particularly important. I found an alternative way of testing this, by introducing a helper method and using a mock (as described [here](https://github.com/enzymejs/enzyme/issues/393#issuecomment-342630403)), but this seems like overkill to me.

What are your thoughts?

File: jest.config.js
Comment (willdurand) [inline jest.config.js:27]: So we can't "just" use the `node` environment? That's what I understood from https://github.com/jsdom/jsdom/issues/2524#issuecomment-902027138

File: tests/unit/blog-utils/TestStaticAddonCard.js
Comment (willdurand) [inline tests/unit/blog-utils/TestStaticAddonCard.js:65]: (I am thinking that these changes in `tests/unit/*` could land ahead of this PR because they will still be useful)

File: jest.config.js
Comment (bobsilverberg) [reply jest.config.js:27]: We can't "just" use the `node` environment, because a large number of our tests (those that use `mount`) expect the DOM to be present. I tried switching to the `node` environment, and then requesting the `jsdom` environment just for a test file that needs it, by adding:
```
/**
 * @jest-environment jsdom
 */
```
to the top of the test file, but that just brings back the ` ReferenceError: TextEncoder is not defined` error. 

I think what we need to do, if we want to make this work, is to not request that the `jsdom` environment be used for these tests, but rather to configure each test to use `jsdom`. I haven't been able to figure out how to do that yet, and even if I do figure it out, it seems like a lot to add to each test, and also then we need to be aware of which test files (or cases) need access to `jsdom` which feels a bit ugly to me.

After looking at all of that, I am of the opinion that we should just accept the "hack" as used in this patch, but if you have a suggestion of another way to avoid it I'm happy to give it a try.

File: tests/unit/blog-utils/TestStaticAddonCard.js
Comment (bobsilverberg) [reply tests/unit/blog-utils/TestStaticAddonCard.js:65]: This is a good idea. I can extract them into a separate issue/PR.

File: jest.config.js
Comment (willdurand) [reply jest.config.js:27]: I agree

Comment (bobsilverberg) [reply jest.config.js:27]: @willdurand As you agree with the above, and you also agreed with removing the one test that no longer seems to work, I think this is good to merge. Would you be able to officially approve the PR, if you are in agreement?

File: tests/unit/blog-utils/TestStaticAddonCard.js
Comment (bobsilverberg) [reply tests/unit/blog-utils/TestStaticAddonCard.js:65]: I think this is actually good to merge now, so I'm not sure it's worthwhile extracting these changes. 

@willdurand If you're fine with this PR now please approve it. If you think it's important that the test changes be extracted let me know and I'll do that in a separate patch.

**Comment by bobsilverberg** [comment]
Ugh, this is unfortunately now failing the `test-next` job. I have re-reun it several times and keep seeing the same thing, which is that output just stops during the test run and then the job fails with "Too long with no output (exceeded 10m0s): context deadline exceeded". I tried to reproduce locally by changing my node to `v16.13.1`, but I couldn't reproduce it. The tests pass.

I've added a bit of debugging code to the `test-ci` step and hopefully that will yield something usable.

**Comment by bobsilverberg** [comment]
Well, this is very strange. It seems that a change I made to the way Jest runs the tests, which I did for debugging purposes, also fixes the tests on node `v16`. The `-i` option, which is shorthand for the `--runInBand` option, causes the tests to run sequentially, instead of in parallel. I thought that forcing them to run that way might make it easier to figure out what was causing the task to timeout on node `v16`, but instead, it simply allows the task to run successfully to completion.

Running Jest with this option for the `test-next` task allows it to pass. It also takes a lot longer to run, but that's a tradeoff I am willing to accept at this point. It would be good, however to try to understand why this is, and if there's another way to address this without using the `-i` option.

I have submitted another test run where I have removed the request for coverage to see if that makes any difference.

**Comment by bobsilverberg** [comment]
> Well, this is very strange. It seems that a change I made to the way Jest runs the tests, which I did for debugging purposes, also fixes the tests on node `v16`. The `-i` option, which is shorthand for the `--runInBand` option, causes the tests to run sequentially, instead of in parallel. I thought that forcing them to run that way might make it easier to figure out what was causing the task to timeout on node `v16`, but instead, it simply allows the task to run successfully to completion.
> 
> Running Jest with this option for the `test-next` task allows it to pass. It also takes a lot longer to run, but that's a tradeoff I am willing to accept at this point. It would be good, however to try to understand why this is, and if there's another way to address this without using the `-i` option.
> 
> I have submitted another test run where I have removed the request for coverage to see if that makes any difference.

The oddness continues. It seems that node `v16` can successfully run the tests if *either* we do not ask for test coverage, *or* we run the tests sequentially, using the `-i` option. I can't see what the relationship between these two things would be, so it seems like a coincidence that both allow the tests to pass. I will dig in a bit more on Wednesday, but for now I'm kind of stumped. I think it's fine to do either of those things for the `test-next` task, for now, but I also would really like to understand why either of those are needed.

@willdurand and/or @diox if either of you have any ideas about what could possibly be going on here, or what else I might do to debug this situation, I'd be keen to hear about it.

File: jest.config.js
Comment (willdurand) [reply jest.config.js:27]: yep, when it isn't failing anymore :)

**Comment by bobsilverberg** [comment]
The weirdness continues. [`47cd9fa` (#11037)](https://github.com/mozilla/addons-frontend/pull/11037/commits/47cd9fadcf36c20494d56d46df0d4c48990602d2) seemed to pass, even though it wasn't run with `-i` and it requested coverage. The only thing different from that and the last run, which failed, was that I specified `--verbose`.  I guess I'll try once again with `--verbose` added to see if it passes again with that option.

**Comment by bobsilverberg** [comment]
Well, I am at a loss, here. Simply adding the `--verbose` option seems to fix things on node `v16`, but I cannot explain why. It seems like a pretty harmless thing to do, other than the fact that it produces *a lot* of output, but even if tests fail, one still gets a summary at the bottom so I don't think it will interfere with our ability to debug. I would love to understand why we need this, but I have also spent *a lot* of time on this PR, so I am inclined, at this point, to just land the patch with `--verbose` and be done with it. I can configure it so that `--verbose` is only used for `test-next`, but I'm not sure that's necessary.

@willdurand @diox What are your thoughts?

**Comment by bobsilverberg** [comment]
Hmm, although the tests were passing, it seems that the changes I made to `src/amo/purify.js` to mimic what was done in https://github.com/mozilla/addons-blog/pull/345/files didn't actually work. When I try to run the site locally I get errors, so I need to track those down first.

**Comment by bobsilverberg** [comment]
And... it's working again. 😄

File: package.json
Comment (willdurand) [inline package.json:134]: Can we remove that flag now?

Comment (bobsilverberg) [reply package.json:134]: The ‘—verbose’ flag? No, we cannot. It’s what is allowing the tests to pass on node 16. See my comment above.

Comment (willdurand) [reply package.json:134]: That PR seemed to have random red and green builds so I was wondering. It would be good to understand why `--verbose` does change the outcome of the test suite, though. Although harmless, I don't think this is a good solution if we cannot explain it.

Comment (bobsilverberg) [reply package.json:134]: To further clarify, the "It's working again" comment at https://github.com/mozilla/addons-frontend/pull/11037#issuecomment-989295100 was only in reference to the fact that the code I wrote broke SSR, which I commented on at https://github.com/mozilla/addons-frontend/pull/11037#issuecomment-989280606. It was not about the node 16 issue, which thus far I have only been able to fix by doing one of 3 things, either:
- add the `-i` flag
- add the `--verbose` flag
- remove the `--coverage` flag

Comment (bobsilverberg) [reply package.json:134]: > That PR seemed to have random red and green builds so I was wondering. It would be good to understand why `--verbose` does change the outcome of the test suite, though. Although harmless, I don't think this is a good solution if we cannot explain it.

I agree, and mentioned this above, but I'm not sure what I can do at this point. There don't seem to be any other debugging options I can add. One thing I haven't tried, and maybe it will just magically fix the problem, is to update this patch to upgrade to jsdom 19 instead of 18. I'll give that a try and cross my fingers really hard.

**Comment by bobsilverberg** [comment]
OMG, I must have crossed my fingers hard enough. It's not passing without any of those hacks! @willdurand Do you want to take one last look at the changes to `src/amo/window.js` and then approve if you're happy?

**Review by willdurand — APPROVED** [review]
🚢

**Comment by bobsilverberg** [comment]
Ship it quick, before it breaks again!

**Comment by bobsilverberg** [comment]
Ugh, this failed after landing. Same problem we've seen above on node 16. I'd rather not back it out. I wonder if we can implement one of the three hacks described above, and open a follow-up to try to figure out what is wrong and how to fix it.

WDYT @willdurand and @diox ?

**Comment by willdurand** [comment]
> WDYT @willdurand and @diox ?

I'd back it out and then try to understand what's going on. If we land a fix, we might not get back to it.

**Comment by bobsilverberg** [comment]
> > WDYT @willdurand and @diox ?
> 
> I'd back it out and then try to understand what's going on. If we land a fix, we might not get back to it.

True. 😢 I'm afraid that it will never land though.  Oh well, it's been outstanding for a long time already, I guess. I'll back it out.

**Comment by willdurand** [comment]
It would be good if we can reopen this PR. I'd like to spend a tiny bit of time to see if I find anything (which I am not sure really)

**Comment by bobsilverberg** [comment]
> It would be good if we can reopen this PR. I'd like to spend a tiny bit of time to see if I find anything (which I am not sure really)

I think GitHub has removed the option to re-open a PR. I can create a new one, and link back to this one, but unfortunately one will have to refer back to this one to see all the comments.

**Pull request merged**
